### PR TITLE
HIVE-25062 Iceberg: Fix date partition transform insert issue

### PR DESCRIPTION
**Pull request by pvary**
### What changes were proposed in this pull request?
Wrap the Record before calculating partitions

### Why are the changes needed?
Timestamp/Date needs values converted to long/int respectively to calculate partitions

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added new unit tests

### CODE REVIEW

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 1]
```diff
@@ -759,6 +763,108 @@ public void testMultilevelIdentityPartitionedWrite() throws IOException {
...
+        optional(1, "id", Types.LongType.get()),
+        optional(2, "part_field", Types.StringType.get()));
+    PartitionSpec spec = PartitionSpec.builderFor(schemaWithDate).truncate("part_field", 2).build();
+    List<Record> records = TestHelper.RecordsBuilder.newInstance(schemaWithDate)
+        .add(1L, "Part1")
+        .add(2L, "Part2")
+        .add(3L, "Art3")
+        .build();
+    Table table = testTables.createTable(shell, "part_test", schemaWithDate, spec, fileFormat, records);
+    HiveIcebergTestUtils.validateData(table, records, 0);
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:863]: What is the difference between the two `validateData` methods?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java [hunk 2]
```diff
@@ -107,6 +110,8 @@
               PrimitiveObjectInspectorFactory.writableStringObjectInspector
           ));
 
+  private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd hh:mm:ss");
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java:113]: Is this used anywhere?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 1 above)
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:863]: Never mind, I guess the new `validateData` reads the data by SQL query, not java API. But then, do we want to keep run both methods?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java [hunk 3]
```diff
@@ -759,6 +763,108 @@ public void testMultilevelIdentityPartitionedWrite() throws IOException {
...
+
+    HiveIcebergTestUtils.validateData(shell, "part_test", records, "id");
+  }
+
+  @Test
+  public void testBucketTransform() throws IOException {
+    Schema schemaWithDate = new Schema(
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:836]: nit: I guess here and in the truncate transform, this is not a schema with a date anymore

Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:836]: nit: I guess here and in the truncate transform, this is not a schema with a date anymore

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 1 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:863]: I would like to check that the data is correct with the Java API and read through SQL is also important

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java (hunk 2 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java:113]: Not anymore

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 3 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java]: renamed.
Good catch! Thx

Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java]: Renamed.
Thx

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java (hunk 2 above)
Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/HiveIcebergTestUtils.java]: Removed

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java (hunk 1 above)
Comment (marton-bod) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:863]: Maybe name one of them `validateDataWithSQL`?

Comment (pvary) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java:863]: Renamed

**Review by marton-bod — APPROVED** [review]

**Review by lcspinter — APPROVED** [review]

**Pull request merged**